### PR TITLE
[ConstraintSystem] Replace dependent member types with holes when the…

### PR DESCRIPTION
**Pull request by hborla**
… base type doesn't conform to the associatedtype's protocol (only in diagnostic mode).

This allows the solver to find solutions for more cases involving requirement failures for dependent member types without special cases across the solver that check for dependent member types with no type variables.

Resolves: rdar://problem/51519565

### CODE REVIEW

**Comment by hborla** [comment]
@swift-ci please smoke test

File: lib/Sema/ConstraintSystem.cpp [hunk 1]
```diff
@@ -2454,8 +2454,13 @@ Type ConstraintSystem::simplifyTypeImpl(Type type,
         auto *proto = assocType->getProtocol();
         auto conformance = DC->getParentModule()->lookupConformance(
           lookupBaseType, proto);
-        if (!conformance)
+        if (!conformance) {
+          if (shouldAttemptFixes() &&
+              getPhase() == ConstraintSystemPhase::Solving)
```
Comment (xedin) [inline lib/Sema/ConstraintSystem.cpp]: Nit: I think it would be worth it to add a comment about why we do this.

**Review by xedin — APPROVED** [review]
Looks great!

File: lib/Sema/ConstraintSystem.cpp (hunk 1 above)
Comment (hborla) [reply lib/Sema/ConstraintSystem.cpp]: Done!

**Comment by hborla** [comment]
@swift-ci please smoke test

File: lib/Sema/ConstraintSystem.cpp (hunk 1 above)
Comment (xedin) [reply lib/Sema/ConstraintSystem.cpp]: Thank you!

**Comment by hborla** [comment]
@swift-ci please smoke test

**Comment by hborla** [comment]
@swift-ci please smoke test OS X Platform

**Pull request merged**